git/git.go

```diff
@@ -0,0 +1,183 @@
+// with a commit.
+type Signature struct {
+	Name  string
+	Email string
+	When  time.Time
+}
+
+// Commit contains all possible information about a Git commit.
+type Commit struct {
+	// Hash is the SHA1 hash of the commit.
+	Hash Hash
+	// Reference is the original reference of the commit, for example:
+	// 'refs/tags/foo'.
+	Reference string
+	// Author is the original author of the commit.
+	Author Signature
+	// Committer is the one performing the commit, might be different from
+	// Author.
+	Committer Signature
+	// Signature is the PGP signature of the commit.
+	Signature string
+	// Encoded is the encoded commit, without any signature.
+	Encoded []byte
+	// Message is the commit message, contains arbitrary text.
+	Message string
+}
+
+// String returns a string representation of the Commit, composed
+// out the last part of the Reference element, and/or Hash.
+// For example: 'tag-1/a0c14dc8580a23f79bc654faa79c4f62b46c2c22',
+// for a "tag-1" tag.
+func (c *Commit) String() string {
+	if short := strings.SplitAfterN(c.Reference, "/", 3); len(short) == 3 {
+		return fmt.Sprintf("%s/%s", short[2], c.Hash)
+	}
+	return fmt.Sprintf("HEAD/%s", c.Hash)
+}
+
+// Verify the Signature of the commit with the given key rings.
+// It returns the fingerprint of the key the signature was verified
+// with, or an error.
+func (c *Commit) Verify(keyRing ...string) (string, error) {
+	if c.Signature == "" {
+		return "", fmt.Errorf("commit does not have a PGP signature")
+	}
+
+	for _, r := range keyRing {
+		reader := strings.NewReader(r)
+		keyring, err := openpgp.ReadArmoredKeyRing(reader)
+		if err != nil {
+			return "", fmt.Errorf("unable to read armored key ring: %w", err)
+		}
+		signer, err := openpgp.CheckArmoredDetachedSignature(keyring, bytes.NewBuffer(c.Encoded), bytes.NewBufferString(c.Signature), nil)
+		if err == nil && len(signer.PrimaryKey.Fingerprint) >= 20 {
+			return fmt.Sprintf("%X", signer.PrimaryKey.Fingerprint[12:20]), nil
+		}
+	}
+	return "", fmt.Errorf("unable to verify commit with any of the given key rings")
+}
+
+// ShortMessage returns the first 50 characters of a commit subject.
+func (c *Commit) ShortMessage() string {
+	subject := strings.Split(c.Message, "\n")[0]
+	r := []rune(subject)
+	if len(r) > 50 {
+		return fmt.Sprintf("%s...", string(r[0:50]))
+	}
+	return subject
+}
+
+// RepositoryReader knows how to perform local and remote read operations
+// to a Git repository.
+type RepositoryReader interface {
+	// Clone clones a repository from the provided url using the options provided.
+	Clone(ctx context.Context, url string, checkoutOpts CheckoutOptions) (*Commit, error)
+	// IsClean returns whether the working tree is clean.
+	IsClean() (bool, error)
+	// Head returns the hash of the current HEAD of the repo.
+	Head() (string, error)
+	// Path returns the path of the repository.
+	Path() string
+	// Cleanup frees any resources that need to be freed at the end of
+	// the client's lifecycle.
+	Cleanup()
```

Comment:
Maybe `Free()` would be more accurate/better? The description also uses "free" to describe it.
Usually we think of freeing resources. Cleanup sounds more like deleting some files.

Reply:
It was `Free()` initially, but changed it after @hiddeco's suggestion: https://github.com/fluxcd/pkg/pull/291#discussion_r911678770

Reply:
Based on this I had another discussion with Sunny. We feel with the recent changes to the interface names, etc. The following might be better:

```go
type Closer interface{
	Close()
}
```

```go
type DiscardCloser struct{}

func (c *DiscardCloser) Close() {}
```

```go
type Client struct {
	*DiscardCloser
        ...
}